File: missing_letter.rb
# A given string contains all the letters from the alphabet except for one. Return the missing letter.

# NOTE: You must accomplish this in O(n) time. This is also known as linear time.

# Input: “The quick brown box jumps over a lazy dog”
# Output: “f”

# def missing_letter(string)
#   alphabet = {
#     "a" => 0, "b" => 0, "c" => 0, "d" => 0, "e" => 0, "f" => 0, "g" => 0, "h" => 0, "i" => 0, "j" => 0, "k" => 0, "l" => 0, "m" => 0, "n" => 0, "o" => 0, "p" => 0, "q" => 0, "r" => 0, "s" => 0, "t" => 0, "u" => 0, "v" => 0, "w" => 0, "x" => 0, "y" => 0, "z" => 0
#   }
#   string.each_char do |letter|
#     if letter != " "
#       alphabet[letter.downcase] += 1
#     end
#   end
#   alphabet.each do |key, value|
#     if value == 0
#       return key
#     end
#   end
# end

# p missing_letter("The quick brown box jumps over a lazy dog")
# p missing_letter("The QUICK brown box jumps OVER a lazy DOG")

def missing_letter(string)
  # store all encountered letter in a hash table
  letter_hash = {}
  string.each_char do |letter|
    letter_hash[letter] = true
  end
  # return first unseen letter
  alphabet = "abcdefghijklmnopqrstuvwxyz"
  alphabet.each_char do |letter|
    if !letter_hash[letter]
      return letter
    end
  end
end

p missing_letter("The quick brown box jumps over a lazy dog") == "f" # true

File: missing-letter.js
// A given string contains all the letters from the alphabet except for one. Return the missing letter.

// Input: “The quick brown box jumps over a lazy dog”
// Output: “f”

// create an empty hash to store the letter of the string argument
// loop through the string input, for each letter, add the letter to the hash with a value of true
// assign a variable called "alphabet" that is a string containing the alphabet
// loop through the alphabet string, if we find a letter that isn't in the hash, return that letter

function missingLetter(string) {
  let letterHash = {};
  for (let i = 0; i < string.length; i++) {
    letterHash[string[i]] = true;
  }
  let alphabet = "abcdefghijklmnopqrstuvwxyz";
  for (let i = 0; i < alphabet.length; i++) {
    if (!letterHash[alphabet[i]]) {
      return alphabet[i];
    }
  }
}

console.log(missingLetter("the quick brown box jumps over a lazy dog"));
